### PR TITLE
feat(backend): Remove PipelineSpec Template storage from ObjStore responsibilies. Fixes #10509

### LINKED ISSUE #10509: [feature] Store Pipeline IR in database, not object storage
### What feature would you like to see?

Currently the Object Store in KFP is largely used for artifacts, except for one outlier, which is the [Pipeline IR](https://github.com/kubeflow/pipelines/blob/2.0.5/backend/src/apiserver/resource/resource_manager.go#L408-L409). 

I agree with the inline comments that this should be stored in the DB just like everything else that's not an artifact. 

### What is the use case or pain point?

Moving this to be stored in db, removes api server's dependency on the object store, and will make it fore future solutions for different artifact store implementations, without having to worry about api server. 

### Is there a workaround currently?

No

### Anything else? 

There's also archive logging, but this seems delegated to the backend engine (currently argo, but soon tekton as well), I'm not sure what to do about this one.

---

Love this idea? Give it a 👍.

### PR DESCRIPTION
**Pull request by gmfrasca**
**Description of your changes:**
Fixes #10509 

The PipelineSpec definition is currently stored in two places, DB and ObjStore, creating the potential for a competing source-of-truth problem.  The ObjStore copy doesn't appear to actually be used/retrieved anywhere, so remove it from the list of responsibilities

**Checklist:**
- [x] The title for your pull request (PR) should follow our title convention. [Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention).

### CODE REVIEW

File: backend/src/apiserver/resource/resource_manager_test.go [hunk 1]
```diff
@@ -1524,18 +1489,6 @@ func TestDeletePipelineVersion_FileError(t *testing.T) {
...
-	manager.objectStore = &FakeBadObjectStore{}
-
-	// Delete the above pipeline_version.
-	err = manager.DeletePipelineVersion(FakeUUIDOne)
-	assert.NotNil(t, err)
-
-	// Verify the version in deleting status.
-	version, err := manager.pipelineStore.GetPipelineVersionWithStatus(FakeUUIDOne, model.PipelineVersionDeleting)
-	assert.Nil(t, err)
-	assert.NotNil(t, version)
```
Comment (gregsheremeta) [inline backend/src/apiserver/resource/resource_manager_test.go:1538]: it didn't jump out at me why this entire block is being deleted. Is there nothing here test-worthy if the object store storage goes away? Specifically wondering if it makes sense to keep the check of the deleting status.

File: backend/src/apiserver/server/fakes_test.go [hunk 2]
```diff
@@ -271,8 +271,9 @@ func initWithExperimentsAndTwoPipelineVersions(t *testing.T) *resource.FakeClien
 	resourceManager = resource.NewResourceManager(clientManager, &resource.ResourceManagerOptions{CollectMetrics: false})
 	_, err = resourceManager.CreatePipelineVersion(
 		&model.PipelineVersion{
-			Name:       "pipeline_version",
-			PipelineId: DefaultFakeUUID,
+			Name:         "pipeline_version",
+			PipelineId:   DefaultFakeUUID,
+			PipelineSpec: "apiVersion: argoproj.io/v1alpha1\nkind: Workflow",
```
Comment (gregsheremeta) [inline backend/src/apiserver/server/fakes_test.go:276]: how is this related to this PR?

Comment (gmfrasca) [reply backend/src/apiserver/server/fakes_test.go:276]: If a mock PipelineSpec isn't provided, the Mock ResourceManager will attempt to fetch the spec from ObjectStore which will break as it no longer is uploaded there.  This behavior only exists in the test RM, I believe

File: backend/src/apiserver/resource/resource_manager_test.go (hunk 1 above)
Comment (gmfrasca) [reply backend/src/apiserver/resource/resource_manager_test.go:1538]: yeah, this specific test can probably remain, albeit the expected results are now the exact opposite: the RM should *not* throw an error if the ObjStore is bad, *should* be able to delete the pipelineversion, and the pipelineversion should *not* exist after performing the delete operation (essentially flip the asserts in L1533, L1537, and L1538)

**Comment by gregsheremeta** [comment]
what's the status of this? Just need reviews?

**Comment by gmfrasca** [comment]
> what's the status of this? Just need reviews?

yes, believe so

**Review by hbelmiro — COMMENTED** [review]
/lgtm

File: backend/src/apiserver/server/fakes_test.go (hunk 2 above)
Comment (HumairAK) [reply backend/src/apiserver/server/fakes_test.go:276]: yep: https://github.com/kubeflow/pipelines/blob/fbb8b399b32e49f0683d2adf85d89dcbc30e3a31/backend/src/apiserver/storage/pipeline_store.go#L433 this is now fetched from the PipelineSpec column in the DB.

**Comment by HumairAK** [comment]
Confirmed this with pre-existing pipelines, they continue to be fetched successfully. New pipelines as well, and they are not stored in the object store. Great work! 

/lgtm
/approve

**Comment by HumairAK** [comment]
@gmfrasca this now has decoupled api server from the object store, can we update the docs here: 

https://github.com/kubeflow/website/blob/master/content/en/docs/components/pipelines/operator-guides/configure-object-store.md

I think we can basically remove the entire first section


fyi @chensun
